### ISSUE TITLE
Fine-tuning SAM2 model with Eval dataset

### ISSUE THREAD

**Issue by zetyquickly** [facebookresearch/sam2#624]
Hello,

Thanks for sharing scripts for SAM model fine-tuning. The usage example is extremely helpful.

However, there's one thing which seems a little bit off: when I try to use `val` by simply copying the`train` part in the config [here](https://github.com/facebookresearch/sam2/blob/main/sam2/configs/sam2.1_training/sam2.1_hiera_b%2B_MOSE_finetune.yaml#L203), I encounter an error at the time of assertion ([here](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/training/trainer.py#L971)), because `set(["all"])` is being excluded from the loss keys. What's the intended way of adding an evaluation stage to the training?

My updated config looks like this:

```

trainer:
...
  mode: train
...

data:
train:
   ...
val:
      _target_: training.dataset.sam2_datasets.TorchTrainMixedDataset
      batch_sizes:
        - ${scratch.train_batch_size}

      datasets:
        - _target_: training.dataset.utils.RepeatFactorWrapper
          dataset:
            _target_: training.dataset.utils.ConcatDataset
            datasets:
            - _target_: training.dataset.vos_dataset.VOSDataset
              transforms: ${vos.train_transforms}
              training: false
              video_dataset:
                _target_: training.dataset.vos_raw_dataset.PNGRawDataset
                img_folder: ${dataset.img_folder}
                gt_folder: ${dataset.gt_folder}
                file_list_txt: ${dataset.val_file_list_txt}
              sampler:
                _target_: training.dataset.vos_sampler.RandomUniformSampler
                num_frames: ${scratch.num_frames}
                max_num_objects: ${scratch.max_num_objects}
              multiplier: ${dataset.multiplier}
      num_workers: ${scratch.num_train_workers}
      pin_memory: True
      shuffle: False
      drop_last: False
      collate_fn:
        _target_: training.utils.data_utils.collate_fn
        _partial_: true
        **dict_key: all**

```

Should the loss and collate_fn be different for the eval stage?

**Comment by zetyquickly** [comment]
@haithamkhedr could you please give a clue of how to properly run the evaluation loop?

**Comment by zetyquickly** [comment]
It feels like removing the whole section [here](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/training/trainer.py#L971) might be a good idea. Maybe I'm wrong

**Comment by ccl-private** [comment]
I have the same need. Any progress?

**Comment by ccl-private** [comment]
by the way, I also want to lock image encoder‘s weights. Any suggestions?

**Comment by zetyquickly** [comment]
comment out the following assert:
```
            if self.loss_conf is not None:
                loss_keys = set(self.loss_conf.keys()) - set(["all"])
                assert all([k in loss_keys for k in val_keys]), (
                    f"Keys in val datasets do not match the keys in losses."
                    f"\nMissing in losses: {set(val_keys) - loss_keys}"
                    f"\nMissing in val datasets: {loss_keys - set(val_keys)}"
                )
```
it's pretty unnecessary to run eval, the similar way as for train

**Comment by zetyquickly** [comment]
to lock the image encoder's weights: It's quite off-topic.

I suggest to use `layer.eval()` + [`layer.requires_grad_` ](https://pytorch.org/docs/stable/generated/torch.Tensor.requires_grad_.html)

**Comment by zetyquickly** [comment]
@NielsRogge could you please have a look, whether it worth a PR. to remove this evaluation blocking code?

> comment out the following assert:
> 
> ```
>             if self.loss_conf is not None:
>                 loss_keys = set(self.loss_conf.keys()) - set(["all"])
>                 assert all([k in loss_keys for k in val_keys]), (
>                     f"Keys in val datasets do not match the keys in losses."
>                     f"\nMissing in losses: {set(val_keys) - loss_keys}"
>                     f"\nMissing in val datasets: {loss_keys - set(val_keys)}"
>                 )
> ```
> 
> it's pretty unnecessary to run eval, the similar way as for train
